lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -619,18 +619,58 @@ MergeVectorizeOps::matchAndRewrite(VectorizeOp vecOp,
     // Ensure that the input vector matches the output of the `otherVecOp`
     // Make sure that the results of the otherVecOp have only one use
     auto otherVecOp = inputVec[0].getDefiningOp<VectorizeOp>();
-    if (!otherVecOp || inputVec != otherVecOp.getResults() ||
-        otherVecOp == vecOp ||
+    if (!otherVecOp || otherVecOp == vecOp ||
         !llvm::all_of(otherVecOp.getResults(),
-                      [](auto result) { return result.hasOneUse(); })) {
+                      [](auto result) { return result.hasOneUse(); }) ||
+        !llvm::all_of(inputVec, [&](auto result) {
+          return result.template getDefiningOp<VectorizeOp>() == otherVecOp;
+        })) {
       newOperands.insert(newOperands.end(), inputVec.begin(), inputVec.end());
       continue;
     }
+
+    // Here, all elements are from the same `VectorizeOp`.
+    // If all elements of the input vector come from the same `VectorizeOp`
+    // sort the vectors by their indices
+    DenseMap<Value, size_t> resultIdxMap;
+    for (auto [resulrIdx, result] : llvm::enumerate(otherVecOp.getResults()))
+      resultIdxMap[result] = resulrIdx;
+
+    SmallVector<Value> tempVec(inputVec.begin(), inputVec.end());
+    llvm::sort(tempVec, [&](Value a, Value b) {
+      return resultIdxMap[a] < resultIdxMap[b];
+    });
+
+    // Check if inputVec matches the result after sorting.
+    if (tempVec != SmallVector<Value>(otherVecOp.getResults().begin(),
+                                      otherVecOp.getResults().end())) {
```

Comment:
Does `tempVec != otherVecOp.getResults()` not work here?

Reply:
I will try, but I think `getResults()` returns `result_range` which cannot be converted into a `SmallVector<Value>` but is worth trying!

Reply:
There should be an implicit conversion from `SmallVector<Value>` to an iterator range, which should theoretically be able to compare against a result range. Worth trying, don't worry if it doesn't work :smiley:

Reply:
I didn't work!
```
 error: invalid operands to binary expression ('SmallVector<mlir::Value>' and '::mlir::Operation::result_range' (aka 'mlir::ResultRange'))
    if (tempVec != otherVecOp.getResults()) 
```

Reply:
Thanks for checking :smiley:

Reply:
@fabianschuiki, should I land this now?

Reply:
It's 12:17 AM here, so please land it if it's Okay.